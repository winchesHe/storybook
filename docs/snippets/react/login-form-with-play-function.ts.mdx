```ts
// LoginForm.stories.ts|tsx

import type { Meta, StoryObj } from '@storybook/react';

import { userEvent, within, expect } from '@storybook/test';

import { LoginForm } from './LoginForm';

const meta: Meta<typeof LoginForm> = {
  component: LoginForm,
};

export default meta;
type Story = StoryObj<typeof LoginForm>;

export const EmptyForm: Story = {};

/*
 * See https://storybook.js.org/docs/writing-stories/play-function#working-with-the-canvas
 * to learn more about using the canvasElement to query the DOM
 */
export const FilledForm: Story = {
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    // ðŸ‘‡ Simulate interactions with the component
    await userEvent.type(canvas.getByTestId('email'), 'email@provider.com');

    await userEvent.type(canvas.getByTestId('password'), 'a-random-password');

    // See https://storybook.js.org/docs/essentials/actions#automatically-matching-args to learn how to setup logging in the Actions panel
    await userEvent.click(canvas.getByRole('button'));

    // ðŸ‘‡ Assert DOM structure
    await expect(
      canvas.getByText(
        'Everything is perfect. Your account is ready and we should probably get you started!',
      ),
    ).toBeInTheDocument();
  },
};
```
