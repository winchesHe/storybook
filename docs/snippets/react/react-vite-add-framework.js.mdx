```js
// .storybook/main.js
export default {
  // ...
  // framework: '@storybook/react-webpack5', ðŸ‘ˆ Remove this
  framework: '@storybook/react-vite', // ðŸ‘ˆ Add this
};
```
