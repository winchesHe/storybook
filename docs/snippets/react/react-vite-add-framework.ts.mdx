```ts
// .storybook/main.ts
import { StorybookConfig } from '@storybook/react-vite';

const config: StorybookConfig = {
  // ...
  // framework: '@storybook/react-webpack5', ðŸ‘ˆ Remove this
  framework: '@storybook/react-vite', // ðŸ‘ˆ Add this
};

export default config;
```
