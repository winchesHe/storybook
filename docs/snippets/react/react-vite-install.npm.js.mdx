```shell
npm install --save-dev @storybook/react-vite
```
