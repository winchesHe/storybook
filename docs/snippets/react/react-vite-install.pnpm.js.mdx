```shell
pnpm install --save-dev @storybook/react-vite
```
