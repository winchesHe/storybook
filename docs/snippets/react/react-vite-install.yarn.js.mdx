```shell
yarn add --dev @storybook/react-vite
```
