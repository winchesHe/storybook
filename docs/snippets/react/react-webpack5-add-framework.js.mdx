```js
// .storybook/main.js
export default {
  // ...
  framework: '@storybook/react-webpack5', // ðŸ‘ˆ Add this
};
```
