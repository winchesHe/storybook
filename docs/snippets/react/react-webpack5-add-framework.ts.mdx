```ts
// .storybook/main.ts
import { StorybookConfig } from '@storybook/react-webpack5';

const config: StorybookConfig = {
  // ...
  framework: '@storybook/react-webpack5', // ðŸ‘ˆ Add this
};

export default config;
```
