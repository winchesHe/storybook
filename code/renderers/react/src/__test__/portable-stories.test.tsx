import React from 'react';
import { vi, it, expect, afterEach, describe } from 'vitest';
import { render, screen, cleanup } from '@testing-library/react';
import { addons } from '@storybook/preview-api';
import type { Meta } from '@storybook/react';
import { expectTypeOf } from 'expect-type';

import { setProjectAnnotations, composeStories, composeStory } from '..';
import type { Button } from './Button';
import * as stories from './Button.stories';

// example with composeStories, returns an object with all stories composed with args/decorators
const { CSF3Primary } = composeStories(stories);

// example with composeStory, returns a single story composed with args/decorators
const Secondary = composeStory(stories.CSF2Secondary, stories.default);
describe('renders', () => {
  afterEach(() => {
    cleanup();
  });

  it('renders primary button', () => {
    render(<CSF3Primary>Hello world</CSF3Primary>);
    const buttonElement = screen.getByText(/Hello world/i);
    expect(buttonElement).not.toBeNull();
  });

  it('reuses args from composed story', () => {
    render(<Secondary />);
    const buttonElement = screen.getByRole('button');
    expect(buttonElement.textContent).toEqual(Secondary.args.children);
  });

  it('onclick handler is called', async () => {
    const onClickSpy = vi.fn();
    render(<Secondary onClick={onClickSpy} />);
    const buttonElement = screen.getByRole('button');
    buttonElement.click();
    expect(onClickSpy).toHaveBeenCalled();
  });

  it('reuses args from composeStories', () => {
    const { getByText } = render(<CSF3Primary />);
    const buttonElement = getByText(/foo/i);
    expect(buttonElement).not.toBeNull();
  });
});

describe('projectAnnotations', () => {
  afterEach(() => {
    cleanup();
  });

  it('renders with default projectAnnotations', () => {
    const WithEnglishText = composeStory(stories.CSF2StoryWithLocale, stories.default);
    const { getByText } = render(<WithEnglishText />);
    const buttonElement = getByText('Hello!');
    expect(buttonElement).not.toBeNull();
  });

  it('renders with custom globals from projectAnnotations via composeStory params', () => {
    const WithPortugueseText = composeStory(stories.CSF2StoryWithLocale, stories.default, {
      globalTypes: { locale: { defaultValue: 'pt' } } as any,
    });
    const { getByText } = render(<WithPortugueseText />);
    const buttonElement = getByText('OlÃ¡!');
    expect(buttonElement).not.toBeNull();
  });

  it('renders with custom projectAnnotations via setProjectAnnotations', () => {
    setProjectAnnotations([{ parameters: { injected: true } }]);
    const Story = composeStory(stories.CSF2StoryWithLocale, stories.default);
    expect(Story.parameters?.injected).toBe(true);
  });
});

describe('CSF3', () => {
  afterEach(() => {
    cleanup();
  });

  it('renders with inferred globalRender', () => {
    const Primary = composeStory(stories.CSF3Button, stories.default);

    render(<Primary>Hello world</Primary>);
    const buttonElement = screen.getByText(/Hello world/i);
    expect(buttonElement).not.toBeNull();
  });

  it('renders with custom render function', () => {
    const Primary = composeStory(stories.CSF3ButtonWithRender, stories.default);

    render(<Primary />);
    expect(screen.getByTestId('custom-render')).not.toBeNull();
  });

  it('renders with play function without canvas element', async () => {
    const CSF3InputFieldFilled = composeStory(stories.CSF3InputFieldFilled, stories.default);

    render(<CSF3InputFieldFilled />);

    await CSF3InputFieldFilled.play!();

    const input = screen.getByTestId('input') as HTMLInputElement;
    expect(input.value).toEqual('Hello world!');
  });

  it('renders with play function with canvas element', async () => {
    const CSF3InputFieldFilled = composeStory(stories.CSF3InputFieldFilled, stories.default);

    const { container } = render(<CSF3InputFieldFilled />);

    await CSF3InputFieldFilled.play!({ canvasElement: container });

    const input = screen.getByTestId('input') as HTMLInputElement;
    expect(input.value).toEqual('Hello world!');
  });
});

// common in addons that need to communicate between manager and preview
it('should pass with decorators that need addons channel', () => {
  const PrimaryWithChannels = composeStory(stories.CSF3Primary, stories.default, {
    decorators: [
      (StoryFn: any) => {
        addons.getChannel();
        return StoryFn();
      },
    ],
  });
  render(<PrimaryWithChannels>Hello world</PrimaryWithChannels>);
  const buttonElement = screen.getByText(/Hello world/i);
  expect(buttonElement).not.toBeNull();
});

describe('ComposeStories types', () => {
  // this file tests Typescript types that's why there are no assertions
  it('Should support typescript operators', () => {
    type ComposeStoriesParam = Parameters<typeof composeStories>[0];

    expectTypeOf({
      ...stories,
      default: stories.default as Meta<typeof Button>,
    }).toMatchTypeOf<ComposeStoriesParam>();

    expectTypeOf({
      ...stories,
      default: stories.default satisfies Meta<typeof Button>,
    }).toMatchTypeOf<ComposeStoriesParam>();
  });
});

// Batch snapshot testing
const testCases = Object.values(composeStories(stories)).map(
  (Story) => [Story.storyName, Story] as [string, typeof Story]
);
it.each(testCases)('Renders %s story', async (_storyName, Story) => {
  cleanup();

  if (_storyName === 'CSF2WithLocale') {
    return;
  }

  const { baseElement } = await render(<Story />);
  await Story.play?.();
  expect(baseElement).toMatchSnapshot();
});
