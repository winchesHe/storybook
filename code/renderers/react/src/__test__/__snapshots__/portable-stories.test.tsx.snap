// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders CSF2Secondary story 1`] = `
<body>
  <div>
    <div
      data-story="true"
      id="storybook-story-example-button--csf-2-secondary"
    >
      <button
        class="storybook-button storybook-button--medium storybook-button--secondary"
        type="button"
      >
        Children coming from story args!
      </button>
    </div>
  </div>
</body>
`;

exports[`Renders CSF2StoryWithLocale story 1`] = `
<body>
  <div>
    <div
      data-story="true"
      id="storybook-story-example-button--csf-2-story-with-locale"
    >
      <button
        class="storybook-button storybook-button--medium storybook-button--secondary"
        type="button"
      >
        Hello!
      </button>
    </div>
  </div>
</body>
`;

exports[`Renders CSF2StoryWithParamsAndDecorator story 1`] = `
<body>
  <div>
    <div
      data-story="true"
      id="storybook-story-example-button--csf-2-story-with-params-and-decorator"
    >
      <button
        class="storybook-button storybook-button--medium storybook-button--secondary"
        type="button"
      >
        foo
      </button>
    </div>
  </div>
</body>
`;

exports[`Renders CSF3Button story 1`] = `
<body>
  <div>
    <div
      data-story="true"
      id="storybook-story-example-button--csf-3-button"
    >
      <button
        class="storybook-button storybook-button--medium storybook-button--secondary"
        type="button"
      >
        foo
      </button>
    </div>
  </div>
</body>
`;

exports[`Renders CSF3ButtonWithRender story 1`] = `
<body>
  <div>
    <div
      data-story="true"
      id="storybook-story-example-button--csf-3-button-with-render"
    >
      <div>
        <p
          data-testid="custom-render"
        >
          I am a custom render function
        </p>
        <button
          class="storybook-button storybook-button--medium storybook-button--secondary"
          type="button"
        >
          foo
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`Renders CSF3InputFieldFilled story 1`] = `
<body>
  <div>
    <div
      data-story="true"
      id="storybook-story-example-button--csf-3-input-field-filled"
    >
      <input
        data-testid="input"
      />
    </div>
  </div>
</body>
`;

exports[`Renders CSF3Primary story 1`] = `
<body>
  <div>
    <div
      data-story="true"
      id="storybook-story-example-button--csf-3-primary"
    >
      <button
        class="storybook-button storybook-button--large storybook-button--primary"
        type="button"
      >
        foo
      </button>
    </div>
  </div>
</body>
`;
